### PR TITLE
[GraphBolt] modify `preprocess_ondisk_dataset()`

### PR DESCRIPTION
**Pull request by Skeleton003**
## Description

#5820 

## Checklist
Please feel free to remove inapplicable items for your PR.
- [ ] The PR title starts with [$CATEGORY] (such as [NN], [Model], [Doc], [Feature]])
- [ ] I've leverage the [tools](https://docs.google.com/document/d/1iHyj7zlmygKSk5gBPsqIqL5ASPzJSPREaNT_QdsiYA4/edit) to beautify the python and c++ code.
- [ ] The PR is complete and small, read the [Google eng practice (CL equals to PR)](https://google.github.io/eng-practices/review/developer/small-cls.html) to understand more about small PR. In DGL, we consider PRs with less than 200 lines of core code change are small (example, test and documentation could be exempted).
- [ ] All changes have test coverage
- [ ] Code is well-documented
- [ ] To the best of my knowledge, examples are either not affected by this change, or have been fixed to be compatible with this change
- [ ] Related issue is referred in this PR
- [ ] If the PR is for a new model/paper, I've updated the example index [here](../examples/README.md).

### CODE REVIEW

**Comment by Skeleton003** [comment]
@dgl-bot

**Comment by Skeleton003** [comment]
plz take a look at your convenience @frozenbugs @czkkkkkk @Rhett-Ying

File: python/dgl/graphbolt/impl/ondisk_dataset.py [hunk 1]
```diff
@@ -115,87 +118,192 @@ def preprocess_ondisk_dataset(
...
         raise RuntimeError("Invalid config: does not contain graph field.")
     is_homogeneous = "type" not in input_config["graph"]["nodes"][0]
     if is_homogeneous:
         # Homogeneous graph.
-        num_nodes = input_config["graph"]["nodes"][0]["num"]
         edge_fmt = input_config["graph"]["edges"][0]["format"]
         edge_path = input_config["graph"]["edges"][0]["path"]
         src, dst = read_edges(dataset_dir, edge_fmt, edge_path)
-        g = dgl.graph((src, dst), num_nodes=num_nodes)
+        num_nodes = input_config["graph"]["nodes"][0]["num"]
```
Comment (Rhett-Ying) [inline python/dgl/graphbolt/impl/ondisk_dataset.py:130]: this change makes the method body too large. pls wrap with function.

Comment (Skeleton003) [reply python/dgl/graphbolt/impl/ondisk_dataset.py]: any suggestions on https://github.com/dmlc/dgl/pull/6986/files#diff-d849f93b0af0c2a14419ea91a6c12bf81851c01425879f1cab7ae058c3fa2278R44 ?

**Comment by Rhett-Ying** [comment]
@Skeleton003  pls resolve the conflicts.

File: python/dgl/graphbolt/impl/ondisk_dataset.py [hunk 2]
```diff
@@ -38,6 +41,215 @@
...
+        The raw data read from yaml file.
+    include_original_edge_id : bool
+        Whether to include the original edge id in the FusedCSCSamplingGraph.
+
+    Returns
+    -------
+    sampling_graph : FusedCSCSamplingGraph
+        The FusedCSCSamplingGraph constructed from the raw data.
+    """
+    is_homogeneous = "type" not in graph_data["nodes"][0]
```
Comment (Rhett-Ying) [inline python/dgl/graphbolt/impl/ondisk_dataset.py:65]: replace with the logic in https://github.com/dmlc/dgl/pull/7011

File: python/dgl/graphbolt/impl/ondisk_dataset.py [hunk 3]
```diff
@@ -38,6 +41,215 @@
 __all__ = ["OnDiskDataset", "preprocess_ondisk_dataset", "BuiltinDataset"]
 
 
+def _graph_data_to_fused_csc_sampling_graph(
```
Comment (Rhett-Ying) [inline python/dgl/graphbolt/impl/ondisk_dataset.py:44]: is it possible to split this func into smaller ones?

File: python/dgl/graphbolt/impl/ondisk_dataset.py [hunk 4]
```diff
@@ -38,6 +41,215 @@
...
+                    f"Edge feature {feat_name} does not cover all edge types. "
+                    f"Existing types: {existing_edge_type}. "
+                    f"Expected types: {all_edge_types}."
+                )
+
+            for feat_name, feat_data in node_feature_collector.items():
+                feat_tensor = torch.empty(
+                    (
+                        [node_type_offset[-1]]
+                        + list(next(iter(feat_data.values())).shape[1:])
```
Comment (Rhett-Ying) [inline python/dgl/graphbolt/impl/ondisk_dataset.py:211]: is this safe? we need to make sure feature data are merged according to `ntype` order, namely the ids of `feat_data.keys()` should be `0~N-1`. please add sanity check logic here to ensure it.

File: python/dgl/graphbolt/impl/ondisk_dataset.py [hunk 5]
```diff
@@ -38,6 +41,215 @@
...
+                    )
+                )
+                for ntype, feat in feat_data.items():
+                    feat_tensor[
+                        node_type_offset[
+                            node_type_to_id[ntype]
+                        ] : node_type_offset[node_type_to_id[ntype] + 1]
+                    ] = feat
+                node_attributes[feat_name] = feat_tensor
+            for feat_name, feat_data in edge_feature_collector.items():
```
Comment (Rhett-Ying) [inline python/dgl/graphbolt/impl/ondisk_dataset.py:221]: ditto. add sanity check here

File: python/dgl/graphbolt/impl/ondisk_dataset.py [hunk 6]
```diff
@@ -38,6 +41,215 @@
...
+                        + list(next(iter(feat_data.values())).shape[1:])
+                    )
+                )
+                for etype, feat in feat_data.items():
+                    feat_tensor[
+                        edge_type_offset[
+                            edge_type_to_id[etype]
+                        ] : edge_type_offset[edge_type_to_id[etype] + 1]
+                    ] = feat
+                edge_attributes[feat_name] = torch.index_select(
```
Comment (Rhett-Ying) [inline python/dgl/graphbolt/impl/ondisk_dataset.py:234]: I am not sure whether we have existing testcase for `node/edge_atrributes`. could you confirm it. If not, please add ones.

File: python/dgl/graphbolt/impl/ondisk_dataset.py [hunk 7]
```diff
@@ -115,109 +327,35 @@ def preprocess_ondisk_dataset(
...
+        sampling_graph,
         os.path.join(
             dataset_dir,
             output_config["graph_topology"]["path"],
         ),
     )
     del output_config["graph"]
 
-    # 7. Load the node/edge features and do necessary conversion.
+    # 5. Load the node/edge features and do necessary conversion.
```
Comment (Rhett-Ying) [inline python/dgl/graphbolt/impl/ondisk_dataset.py:358]: this reminds me that do we need to shuffle edge feature data that are saved into FeatureStore. are the edge IDs remain the same after convert raw graph data to FusedCSCSamplingGraph? any testcases for this?

File: python/dgl/graphbolt/impl/ondisk_dataset.py [hunk 8]
```diff
@@ -38,6 +41,215 @@
...
+        )
+        indptr, indices, value_indices = sparse_matrix.csc()
+        node_type_offset = torch.tensor(node_type_offset)
+        type_per_edge = torch.index_select(
+            coo_etype, dim=0, index=value_indices
+        )
+        node_attributes = {}
+        edge_attributes = {}
+        if include_original_edge_id:
+            edge_attributes[ORIGINAL_EDGE_ID] = value_indices
```
Comment (Rhett-Ying) [inline python/dgl/graphbolt/impl/ondisk_dataset.py:127]: original EIDs are supposed to be per-type IDs. does `value_indices` are per-type or global EIDs?

File: python/dgl/graphbolt/impl/ondisk_dataset.py (hunk 6 above)
Comment (Skeleton003) [reply python/dgl/graphbolt/impl/ondisk_dataset.py:234]: yes, but only tested whether some specific feature name is added into node/edge_atrributes because we use random graph and features in testcases. I shall add hard-code testcases for preprocessing by the way.

File: python/dgl/graphbolt/impl/ondisk_dataset.py (hunk 8 above)
Comment (Skeleton003) [reply python/dgl/graphbolt/impl/ondisk_dataset.py:127]: Yes it is, and dedicated tests will be added thanks to your remind

File: python/dgl/graphbolt/impl/ondisk_dataset.py (hunk 3 above)
Comment (Skeleton003) [reply python/dgl/graphbolt/impl/ondisk_dataset.py:44]: I'm afraid not.

File: python/dgl/graphbolt/impl/ondisk_dataset.py [hunk 9]
```diff
@@ -38,6 +41,223 @@
...
+    sampling_graph : FusedCSCSamplingGraph
+        The FusedCSCSamplingGraph constructed from the raw data.
+    """
+    is_homogeneous = (
+        len(graph_data["nodes"]) == 1
+        and len(graph_data["edges"]) == 1
+        and "type" not in graph_data["nodes"][0]
+        and "type" not in graph_data["edges"][0]
+    )
+
```
Comment (mfbalin) [inline python/dgl/graphbolt/impl/ondisk_dataset.py:71]: `import dgl.sparse as dglsp`
Would placing the dgl.sparse import here instead of the top of the file help solve the issue?

Comment (mfbalin) [reply python/dgl/graphbolt/impl/ondisk_dataset.py:71]: I think `dgl.sparse` and `dgl.graphbolt` should not depend on each other. We should also make an effort to reduce the dependency of graphbolt to even dgl.

Comment (Skeleton003) [reply python/dgl/graphbolt/impl/ondisk_dataset.py:71]: I see. Thanks for you kind advice!

Comment (mfbalin) [reply python/dgl/graphbolt/impl/ondisk_dataset.py:71]: Looks like the cugraph test passed. #7077 and #7075 should not be needed anymore I guess.

Comment (Skeleton003) [reply python/dgl/graphbolt/impl/ondisk_dataset.py:71]: Yes. thank you again

File: python/dgl/graphbolt/impl/ondisk_dataset.py [hunk 10]
```diff
@@ -38,6 +39,234 @@
...
+        node_type_offset = torch.tensor(node_type_offset)
+        type_per_edge = torch.index_select(
+            coo_etype, dim=0, index=value_indices
+        )
+        node_attributes = {}
+        edge_attributes = {}
+        if include_original_edge_id:
+            original_edge_id = [
+                value_indices[i] - edge_type_offset[type_per_edge[i]]
+                for i in range(total_num_edges)
```
Comment (Rhett-Ying) [inline python/dgl/graphbolt/impl/ondisk_dataset.py:141]: explicit for loop could be quite inefficient. equivalent with below one which eliminate explicit loop and converting to tensor back.
```
value_indices - edge_type_offset[type_per_edge[value_indices]]
```

**Comment by Rhett-Ying** [comment]
please check if such peak memory issue happens in current implementation. https://github.com/dmlc/dgl/issues/7086

**Comment by Skeleton003** [comment]
Benchmarking results are pasted in the description. @Rhett-Ying  @czkkkkkk 

Good news is that the new implementation is generally faster than the old one, and is more memory-saving for homo datasets (ogbn-products).

But it is worth noting that the new implementation uses more memory to deal with datasets with a heterograph (ogbn-mag). I am going through the code to figure out why, but I don't have a clue right now. Your inspiration on this issue will be appreciated.

**Comment by Skeleton003** [comment]
Tried to change the dtype of `type_per_edge` for hetero graph:

After change `type_per_edge` to `torch.int32`:
```sh
Version: old | Dataset: ogbn-mag | include_EID: True
Memory Usage: 8.008 GB | Execution Time: 14.99524 seconds
Version: new | Dataset: ogbn-mag | include_EID: True
Memory Usage: 8.682 GB | Execution Time: 13.81134 seconds
 
Version: old | Dataset: ogbn-mag | include_EID: False
Memory Usage: 7.976 GB | Execution Time: 13.61416 seconds
Version: new | Dataset: ogbn-mag | include_EID: False
Memory Usage: 8.053 GB | Execution Time: 12.39217 seconds
```

After change `type_per_edge` to `torch.int16`:
```
Version: old | Dataset: ogbn-mag | include_EID: True
Memory Usage: 8.008 GB | Execution Time: 15.08433 seconds
Version: new | Dataset: ogbn-mag | include_EID: True
Memory Usage: 8.448 GB | Execution Time: 13.50126 seconds
 
Version: old | Dataset: ogbn-mag | include_EID: False
Memory Usage: 7.976 GB | Execution Time: 13.57649 seconds
Version: new | Dataset: ogbn-mag | include_EID: False
Memory Usage: 7.819 GB | Execution Time: 12.00695 seconds
```

**Comment by Rhett-Ying** [comment]
@Skeleton003 the default or previous dtype of `type_per_edge` is `torch.int64`? could you paste the results for `ogbn-products` as well? And how do you measure the memory usage?

**Comment by Rhett-Ying** [comment]
According to the `ogbn-mag` results, storing edge attributes has impact on memory usage.

File: python/dgl/graphbolt/impl/ondisk_dataset.py [hunk 11]
```diff
@@ -38,6 +41,232 @@
...
+        )
+        indptr, indices, value_indices = sparse_matrix.csc()
+        node_type_offset = torch.tensor(node_type_offset)
+        type_per_edge = torch.index_select(
+            coo_etype, dim=0, index=value_indices
+        )
+        node_attributes = {}
+        edge_attributes = {}
+        if include_original_edge_id:
+            edge_attributes[ORIGINAL_EDGE_ID] = value_indices - torch.gather(
```
Comment (Rhett-Ying) [inline python/dgl/graphbolt/impl/ondisk_dataset.py:141]: should this be blamed for memory consumption increase?

File: python/dgl/graphbolt/impl/ondisk_dataset.py [hunk 12]
```diff
@@ -38,6 +41,232 @@
...
+                input=torch.tensor(edge_type_offset), dim=0, index=type_per_edge
+            )
+
+    # Load the sampling related node/edge features and add them to
+    # the sampling-graph.
+    if graph_data.get("feature_data", None):
```
Comment (Rhett-Ying) [inline python/dgl/graphbolt/impl/ondisk_dataset.py:147]: could you measure the memory footprint of this part as well?

File: python/dgl/graphbolt/impl/ondisk_dataset.py (hunk 11 above)
Comment (Skeleton003) [reply python/dgl/graphbolt/impl/ondisk_dataset.py:141]: If so, how to improve it?

**Comment by Skeleton003** [comment]
> @Skeleton003 the default or previous dtype of `type_per_edge` is `torch.int64`? could you paste the results for `ogbn-products` as well? And how do you measure the memory usage?

Yes, default or previous dtype of `type_per_edge` is `torch.int64`. The code and results are pasted in the description.

File: python/dgl/graphbolt/impl/ondisk_dataset.py [hunk 13]
```diff
@@ -38,6 +41,232 @@
...
+        coo_etype = torch.cat(coo_etype_list)
+
+        total_num_nodes = node_type_offset[-1]
+        total_num_edges = edge_type_offset[-1]
+
+        sparse_matrix = spmatrix(
+            indices=torch.stack((coo_src, coo_dst), dim=0),
+            shape=(total_num_nodes, total_num_nodes),
+        )
+        indptr, indices, value_indices = sparse_matrix.csc()
```
Comment (Rhett-Ying) [inline python/dgl/graphbolt/impl/ondisk_dataset.py:133]: after this, could `coo_src`, `coo_dst` can be deleted?

File: python/dgl/graphbolt/impl/ondisk_dataset.py (hunk 11 above)
Comment (Rhett-Ying) [reply python/dgl/graphbolt/impl/ondisk_dataset.py:141]: can `value_indices` be re-used here?
```
value_indices -= torch.gather() # in-place subtraction
```

Comment (Rhett-Ying) [reply python/dgl/graphbolt/impl/ondisk_dataset.py:141]: at least we could re-use the tensor storage of `torch.gather`

Comment (Skeleton003) [reply python/dgl/graphbolt/impl/ondisk_dataset.py:141]: > can `value_indices` be re-used here?
> 
> ```
> value_indices -= torch.gather() # in-place subtraction
> ```

What is the difference? I think their memory consumption are the same.

Comment (Skeleton003) [reply python/dgl/graphbolt/impl/ondisk_dataset.py:141]: Oh it works, amazing.

**Comment by Rhett-Ying** [comment]
So the remaining work items are

- [ ] > profile memory footprint of feature data.
- [ ] > re-preprocess datasets and run end2end example to check if accuracy is expected.(just make sure the new processing works well)

**Comment by Skeleton003** [comment]
Benchmark on self-made dataset with graph-features, containing a heterograph with 3 million nodes and 30 million edges.

```sh
Version: old | Dataset: selfmade | include_EID: True
Memory Usage: 12.004 GB | Execution Time: 26.74319 seconds
Version: new | Dataset: selfmade | include_EID: True
Memory Usage: 10.524 GB | Execution Time: 26.45344 seconds
 
Version: old | Dataset: selfmade | include_EID: False
Memory Usage: 12.004 GB | Execution Time: 25.80169 seconds
Version: new | Dataset: selfmade | include_EID: False
Memory Usage: 10.524 GB | Execution Time: 25.55227 seconds
```

**Comment by Skeleton003** [comment]
- [x]  profile memory footprint of feature data.
- [x]  re-preprocess datasets and run end2end example to check if accuracy is expected.(just make sure the new processing works well)
